plugins/linux-swap/fu-linux-swap.c

```diff
@@ -13,46 +13,256 @@
 
 struct _FuLinuxSwap {
 	GObject		 parent_instance;
-	gboolean	 encrypted;
-	gboolean	 enabled;
+	guint		 encrypted_cnt;
+	guint		 enabled_cnt;
 };
 
 G_DEFINE_TYPE (FuLinuxSwap, fu_linux_swap, G_TYPE_OBJECT)
 
+static gchar *
+fu_strdup_nospaces (const gchar *line)
+{
+	GString *str = g_string_new (NULL);
+	for (guint i = 0; line[i] != '\0' && !g_ascii_isspace (line[i]); i++)
+		g_string_append_c (str, line[i]);
+	return g_string_free (str, FALSE);
+}
+
+#define UDISKS_DBUS_SERVICE		"org.freedesktop.UDisks2"
+#define UDISKS_DBUS_PATH		"/org/freedesktop/UDisks2/Manager"
+#define UDISKS_DBUS_MANAGER_INTERFACE	"org.freedesktop.UDisks2.Manager"
+#define UDISKS_DBUS_BLOCK_INTERFACE	"org.freedesktop.UDisks2.Block"
+
+static GPtrArray *
+fu_common_get_block_devices (GError **error)
+{
+	GVariantBuilder builder;
+	const gchar *obj;
+	g_autoptr(GVariant) output = NULL;
+	g_autoptr(GDBusProxy) proxy = NULL;
+	g_autoptr(GPtrArray) devices = NULL;
+	g_autoptr(GVariantIter) iter = NULL;
+	g_autoptr(GDBusConnection) connection = NULL;
+
+	connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, error);
+	if (connection == NULL) {
+		g_prefix_error (error, "failed to get system bus: ");
+		return FALSE;
+	}
+	proxy = g_dbus_proxy_new_sync (connection,
+				       G_DBUS_PROXY_FLAGS_NONE, NULL,
+				       UDISKS_DBUS_SERVICE,
+				       UDISKS_DBUS_PATH,
+				       UDISKS_DBUS_MANAGER_INTERFACE,
+				       NULL, error);
+	if (proxy == NULL) {
+		g_prefix_error (error, "failed to find %s: ", UDISKS_DBUS_SERVICE);
+		return NULL;
+	}
+	g_variant_builder_init (&builder, G_VARIANT_TYPE_VARDICT);
+	output =  g_dbus_proxy_call_sync (proxy,
+					  "GetBlockDevices",
+					  g_variant_new ("(a{sv})", &builder),
+					  G_DBUS_CALL_FLAGS_NONE,
+					  -1, NULL, error);
+	if (output == NULL)
+		return NULL;
+	devices = g_ptr_array_new_with_free_func (g_free);
+	g_variant_get (output, "(ao)", &iter);
+	while (g_variant_iter_next (iter, "&o", &obj))
+		g_ptr_array_add (devices, g_strdup (obj));
+
+	return g_steal_pointer (&devices);
+}
```

Comment:
this should go into `fu-common.c` where we have already used udisks for siimilar tasks

Reply:
Refactoring now.

Reply:
Okay, how about something like that.